packages/cdktf/lib/terraform-backend.ts

```diff
@@ -37,29 +37,95 @@ export abstract class TerraformBackend extends TerraformElement {
   ): TerraformRemoteState;
 
   public toHclTerraform(): any {
+    const backendConfig = deepMerge(
+      this.synthesizeHclAttributes(),
+      this.rawOverrides
+    );
+
+    const variables = this.detectVariables(backendConfig);
+
+    if (variables.length > 0) {
+      throw new Error(
+        [
+          "Terraform does not support interpolated values within backends.",
+          "Interpolations found in the following configuration backend attributes:",
+          ...variables.map((v) => `\t${this.constructor.name}: "${v}"`),
+        ].join("\n")
+      );
+    }
     return {
       terraform: {
         backend: {
-          [this.name]: deepMerge(
-            this.synthesizeHclAttributes(),
-            this.rawOverrides
-          ),
+          [this.name]: backendConfig,
         },
       },
     };
   }
 
+  private detectVariables(obj: any, parentKey?: string): string[] {
+    const hasVariableRegex = /var\.(\w+)/g;
+
+    if (typeof obj === "string") {
+      const matches = hasVariableRegex.test(obj);
+      if (matches) {
+        if (parentKey) {
+          return [parentKey];
+        }
+      }
+
+      return [];
+    } else if (typeof obj === "object") {
+      // References
+      if (obj.value && typeof obj.value === "string") {
+        return this.detectVariables(obj.value, parentKey);
+      }
+
+      const foundKeys = Object.keys(obj).filter((key) => {
+        return (
+          this.detectVariables(
+            obj[key],
+            parentKey ? `${parentKey}.${key}` : key
+          ).length > 0
+        );
+      });
+
+      return foundKeys.flat(0);
+    } else if (Array.isArray(obj)) {
+      const foundIndices = obj.filter((v, i) =>
+        this.detectVariables(v, `${parentKey}[${i}]`)
+      );
+
+      return foundIndices.flat(0);
+    } else {
+      return [];
+    }
+  }
+
   /**
    * Adds this resource to the terraform JSON output.
    */
   public toTerraform(): any {
+    const backendConfig = deepMerge(
+      this.synthesizeAttributes(),
+      this.rawOverrides
+    );
+
+    const variables = this.detectVariables(backendConfig);
+
+    if (variables.length > 0) {
+      throw new Error(
+        [
+          "Terraform does not support interpolated values within backends.",
+          "Interpolations found in the following configuration backend attributes:",
+          ...variables.map((v) => `\t${this.constructor.name}: "${v}"`),
+        ].join("\n")
+      );
```

Comment:
I'd prefer to add an error annotation instead. This will automatically print the construct id alongside and it won't short circuit after the first error.
Here's how they look for example: https://github.com/hashicorp/terraform-cdk/blob/5627fe99b08e931c32594a2403e05c86d9cc6e91/test/typescript/synth-error-annotations/test.ts#L23